### PR TITLE
Add/material design icons and svg4everybody

### PR DESCRIPTION
**Pull request by lynnbgriffith**
#### Changes proposed in this Pull Request

Add a package for storing the Material icons we're using in the new nav drawer work (see PR #31742). The official material-design-icon package (https://github.com/google/material-design-icons) has not been updated since Jan 2018. It is missing several icons we intend to use.

To load SVGs, I use the technique @sgomes introduced in PR #32172.  This uses svg4everybody to polyfill the use of external references in SVG `<use>` elements, enabling the browser to load and optimize SVG rather than converting SVGs to React components.

#### Testing instructions

With this PR, you can use the new MaterialIcon component to render an icon.  Example:

```jsx
import MaterialIcon from 'components/material-icon';  

const Icon = <MaterialIcon size={ 24 } icon="offline_bolt" />;
```

### CODE REVIEW

File: client/components/material-icon/index.jsx [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+	svg4everybody();
+}
+
+function MaterialIcon( props ) {
+	const { size = 24, style = 'outline', icon, onClick, className, ...otherProps } = props;
+
+	// Using a missing icon doesn't produce any errors, just a blank icon, which is the exact intended behaviour.
+	// This means we don't need to perform any checks on the icon name.
+	const iconName = `material-icon-${ icon }`;
+	const iconClass = [ 'material-icon', iconName, className ]
```
Comment (blowery) [inline client/components/material-icon/index.jsx]: could probably use `classnames` to make this simpler

File: client/components/material-icon/index.jsx [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * External dependencies
+ */
+import React from 'react';
+import PropTypes from 'prop-types';
+import svg4everybody from 'svg4everybody';
+
+const isBrowser = typeof window !== 'undefined';
+if ( isBrowser ) {
```
Comment (blowery) [inline client/components/material-icon/index.jsx]: @sgomes @griffbrad do we want this at every usage or should we move this to `lib/polyfill`?

File: client/components/material-icon/index.jsx [hunk 3]
```diff
@@ -0,0 +1,52 @@
+/** @format */
```
Comment (blowery) [inline client/components/material-icon/index.jsx]: don't need the pragma anymore

File: packages/material-design-icons/package.json [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+	"files": [
+		"src"
+	],
+	"devDependencies": {
+		"grunt": "^0.4.5",
+		"grunt-cli": "^1.2.0",
+		"grunt-svgstore": "^1.0.0"
+	},
+	"scripts": {
+		"build": "grunt --verbose"
```
Comment (blowery) [inline packages/material-design-icons/package.json]: could we do this with the [svgstore cli](https://github.com/svgstore/svgstore-cli) and not pull in grunt?

Comment (griffbrad) [reply packages/material-design-icons/package.json]: Thank you.  Did not stumble across this and just followed the gridicons example.  Very happy to not have to touch grunt.

File: client/components/material-icon/index.jsx [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+		<svg
+			xmlns="http://www.w3.org/2000/svg"
+			viewBox="0 0 24 24"
+			className={ iconClass }
+			height={ size }
+			width={ size }
+			onClick={ onClick }
+			{ ...otherProps }
+		>
+			<use xlinkHref={ `/calypso/images/material-design-icons/material-icons.svg#${ svgPath }` } />
```
Comment (blowery) [inline client/components/material-icon/index.jsx]: this needs to get cache-busted somehow. maybe file-loader or url-loader to get the path?

File: client/components/material-icon/index.jsx (hunk 1 above)
Comment (griffbrad) [reply client/components/material-icon/index.jsx]: Added in fd5fbba.

File: client/components/material-icon/index.jsx (hunk 3 above)
Comment (griffbrad) [reply client/components/material-icon/index.jsx]: Removed in 6573c2c.

File: packages/material-design-icons/package.json (hunk 4 above)
Comment (griffbrad) [reply packages/material-design-icons/package.json]: Made the switch in 426e7a8.

File: package.json [hunk 6]
```diff
@@ -219,7 +221,7 @@
 		"build-devdocs:index": "cross-env-shell NODE_PATH=$NODE_PATH:server:client:. npm run -s build-devdocs:index:_env",
 		"build-devdocs:index:_env": "node server/devdocs/bin/generate-devdocs-index",
 		"build-docker": "node bin/build-docker.js",
-		"build-static": "ncp static public",
+		"build-static": "ncp packages/material-design-icons/svg-sprite public/images/material-design-icons && ncp static public",
```
Comment (blowery) [inline package.json]: this is failing after a `clean`. probably need to copy static into public first, then copy in the icons?

Comment (griffbrad) [reply package.json]: This is removed after the switch to using file-loader in cb90ef8.

File: client/components/material-icon/index.jsx (hunk 5 above)
Comment (griffbrad) [reply client/components/material-icon/index.jsx]: Using file-loader in cb90ef8.

File: webpack.config.js [hunk 7]
```diff
@@ -250,6 +250,7 @@ const webpackConfig = {
 				'gridicons/example': 'gridicons/dist/example',
 				'react-virtualized': 'react-virtualized/dist/es',
 				'social-logos/example': 'social-logos/build/example',
+				'material-design-icons': path.join( __dirname, 'packages/material-design-icons' ),
```
Comment (blowery) [inline webpack.config.js]: i think you could drop this and just import from the package name

File: client/components/material-icon/index.jsx [hunk 8]
```diff
@@ -0,0 +1,49 @@
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
+import React from 'react';
+import PropTypes from 'prop-types';
+import svg4everybody from 'svg4everybody';
+import spritePath from 'material-design-icons/svg-sprite/material-icons.svg';
```
Comment (blowery) [inline client/components/material-icon/index.jsx]: can this be `@automattic/material-design-icons/svg-sprite/material-icons.svg` instead? would remove the need for the alias

File: packages/material-design-icons/package.json [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+		"directory": "packages/material-design-icons"
+	},
+	"publishConfig": {
+		"access": "private"
+	},
+	"bugs": {
+		"url": "https://github.com/Automattic/wp-calypso/issues"
+	},
+	"files": [
+		"src"
```
Comment (blowery) [inline packages/material-design-icons/package.json]: should include svg-sprite here

Comment (griffbrad) [reply packages/material-design-icons/package.json]: Added in ced90eb, which resolves the import alias issue as well.

File: client/components/material-icon/index.jsx (hunk 2 above)
Comment (griffbrad) [reply client/components/material-icon/index.jsx]: I'd favor adding it to client/boot/polyfills.

Comment (griffbrad) [reply client/components/material-icon/index.jsx]: Made this switch in b89d362.  Seems to fit in better with the way we handle other polyfills.

**Comment by blowery** [comment]
@griffbrad I added the MaterialIcon to the playground scope just to be able to test it out there. Looks good! We might want to add an example to devdocs too, but that can probably wait for another PR.

**Review by blowery — APPROVED** [review]
LGTM

File: client/components/material-icon/index.jsx (hunk 2 above)
Comment (sgomes) [reply client/components/material-icon/index.jsx]: Looks good to me 👍

File: packages/material-design-icons/README.md [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+This package currently only provides Material icon SVGs required by the Calypso
+nav drawer.  The official `material-design-icons` package (https://github.com/google/material-design-icons)
+is generally quite out-dated.  It also includes many image formats that are not
+relevant to Calypso.
+
+As we adopt Material icons more widely, it will probably be worthwhile to fully
+populate this package with all the available icons.  Alternatively, we could seek
+out an alternative distribution of the icons.
+
+## Using a Material icon in Calypso
```
Comment (sgomes) [inline packages/material-design-icons/README.md]: @griffbrad Looks like the readme needs to be updated in a future PR.

File: client/components/material-icon/index.jsx [hunk 11]
```diff
@@ -0,0 +1,42 @@
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
+import React from 'react';
+import PropTypes from 'prop-types';
+import spritePath from '@automattic/material-design-icons/svg-sprite/material-icons.svg';
```
Comment (sgomes) [inline client/components/material-icon/index.jsx]: @griffbrad This line is interesting. How does this import get converted into a path string, and how would I be able to replicate it in gridicons?

**Comment by sgomes** [comment]
Excellent work, @griffbrad, some great improvements on my experimental PR! I'll be sure to copy what you did here when I start work on gridicons 🙂

File: packages/material-design-icons/README.md (hunk 10 above)
Comment (griffbrad) [reply packages/material-design-icons/README.md]: Thanks for the heads up.  Will post that later today.

File: client/components/material-icon/index.jsx (hunk 11 above)
Comment (blowery) [reply client/components/material-icon/index.jsx]: https://github.com/Automattic/wp-calypso/blob/master/packages/calypso-build/webpack/file-loader.js

file-loader magic

Comment (griffbrad) [reply client/components/material-icon/index.jsx]: This is done with webpack's file loader.  In the context of Calypso, it's added to webpack.config.js here:
https://github.com/Automattic/wp-calypso/blob/master/webpack.config.js#L234

That comes out of the Calypso build package:
https://github.com/Automattic/wp-calypso/blob/master/packages/calypso-build/webpack/file-loader.js

Comment (sgomes) [reply client/components/material-icon/index.jsx]: Very interesting, thank you both! Should be easy to replicate for gridicons, then :)

**Comment by paaljoachim** [comment]
I am thinking that you might have some good input on the Add icons block being proposed for Gutenberg Core. https://github.com/WordPress/gutenberg/issues/16484

**Comment by sgomes** [comment]
Thanks, @paaljoachim, I'll take a look!

**Pull request merged**